ci/docker/conda-cpp.dockerfile

```diff
@@ -54,6 +56,7 @@ ENV ARROW_BUILD_TESTS=ON \
     ARROW_WITH_SNAPPY=ON \
     ARROW_WITH_ZLIB=ON \
     ARROW_WITH_ZSTD=ON \
+    CMAKE_CXX_STANDARD=17 \
```

Comment:
Note this will break OSX for now, so it should only be set for Linux (as it is done here). See the tracking issue https://github.com/conda-forge/clang-compiler-activation-feedstock/issues/17 for that. Can we remove the explicit setting of `CMAKE_CXX_STANDARD` in CMake instead?

Reply:
@pitrou do you have any thoughts on whether we can change the default for `CMAKE_CXX_STANDARD`?

Reply:
Hmm. I'm not sure I understand the situation entirely, but for now Arrow C++ needs to compile on gcc 4.9 (and perhaps even gcc 4.8, for a subset of Arrow). This is because of the compiler requirements for R packages...

Reply:
Also, unless I misunderstand @xhochy 's first message, it seems there's no real problem here?

Reply:
My message was mainly a heads-up that on OSX, you need to set `CMAKE_CXX_STANDARD` to `14`. This is Linux here, so everything is fine here.

I'm not sure about the general removal of `CMAKE_CXX_STANDARD` as google-cloud-cpp does also set it: https://github.com/googleapis/google-cloud-cpp/blob/13ec1e946ae1baad6bcae952daf5910649dcfd0a/CMakeLists.txt#L31-L41 Possibly that combination could also be used here?

Reply:
Well, we already have the [following](https://github.com/apache/arrow/blob/master/cpp/cmake_modules/SetupCxxFlags.cmake#L122-L124):
```cmake
if(NOT DEFINED CMAKE_CXX_STANDARD)
  set(CMAKE_CXX_STANDARD 11)
endif()
```

The only difference AFAICT is that we don't error out if the user explicitly asked for something earlier than C++11, but that must be a _really_ rare case.

Reply:
@xhochy Can you remind me the reason of `CMAKE_CXX_STANDARD=17` here? Also, perhaps explain it in a comment?